### PR TITLE
[nnyeah] Beginnings of an integration test framework

### PR DESCRIPTION
**Pull request by chamons**
In the future this will use a .NET tool to run against some synthetic test cases and some real nugets.

Right now this is enough to know that things are busted in main:

```
Unable to generate output file, unexpected exception: An item with the same key has already been added. Key: System.Boolean System.nint::System.IConvertible.ToBoolean(System.IFormatProvider).make[1]: [run-tests] Error 1 (ignored)
Unable to generate output file, unexpected exception: An item with the same key has already been added. Key: System.Boolean System.nint::System.IConvertible.ToBoolean(System.IFormatProvider).make[1]: [run-tests] Error 1 (ignored)
Unable to generate output file, unexpected exception: An item with the same key has already been added. Key: System.Boolean System.nint::System.IConvertible.ToBoolean(System.IFormatProvider).make[1]: [run-tests] Error 1 (ignored)
```

My make is a bit rusty, bike shedding is welcome .

### CODE REVIEW

File: tools/nnyeah/integration/Makefile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=ios/bin/Debug/iOSIntegration.dll --output=Converted/iOSIntegration.dll
+	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=tv/bin/Debug/TVIntegration.dll --output=Converted/TVIntegration.dll	
+
+mac/bin/Debug/MacIntegration.dll: $(wildcard *.cs */*.cs Makefile) mac/MacIntegration.csproj
+	$(Q) $(SYSTEM_MSBUILD) mac/MacIntegration.csproj $(MSBUILD_VERBOSITY)
+
+ios/bin/Debug/iOSIntegration.dll: $(wildcard *.cs */*.cs Makefile) ios/iOSIntegration.csproj
+	$(Q_BUILD) $(SYSTEM_MSBUILD) ios/iOSIntegration.csproj $(MSBUILD_VERBOSITY)
+
+tv/bin/Debug/TVIntegration.dll: $(wildcard *.cs */*.cs Makefile) tv/TVIntegration.csproj
```
Comment (stephen-hawley) [inline tools/nnyeah/integration/Makefile:21]: careful on scope creep. AFAIK, we're not doing TVOS.

**Review by stephen-hawley — APPROVED** [review]

File: tools/nnyeah/integration/Makefile (hunk 1 above)
Comment (chamons) [reply tools/nnyeah/integration/Makefile:21]: Good to know. 

I mostly added it because TVOS device tests tend to be more stable/easy to work with than iOS, so even if we're not supporting it officially it might make potential future device tests less painful.

File: tools/nnyeah/integration/Makefile [hunk 2]
```diff
@@ -0,0 +1,27 @@
+TOP=../../..
+include $(TOP)/Make.config
+
+all-local:: mac/bin/Debug/MacIntegration.dll ios/bin/Debug/iOSIntegration.dll tv/bin/Debug/TVIntegration.dll
```
Comment (rolfbjarne) [inline tools/nnyeah/integration/Makefile:4]: It would be nice to use a consistent platform naming: `macOS`, `iOS`, `tvOS`.

That way it's much easier to handle multiple platforms (we have a list of platforms we're building for with that naming), to:

* Not test platforms we're not building for.
* Maybe one day iterate over all the platforms (as opposed to hardcoding them).

File: tools/nnyeah/integration/Makefile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+install-local:: all-local
+
+run-tests::
+	$(Q) rm -rf converted
+	$(Q) mkdir -p converted	
+	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=mac/bin/Debug/MacIntegration.dll --output=Converted/MacIntegration.dll
+	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=ios/bin/Debug/iOSIntegration.dll --output=Converted/iOSIntegration.dll
+	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=tv/bin/Debug/TVIntegration.dll --output=Converted/TVIntegration.dll
```
Comment (rolfbjarne) [inline tools/nnyeah/integration/Makefile:13]: By prepending the dash, you're preventing detection of failures from scripts ("make run-tests" will always succeed, even if there are test failures).

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tools/nnyeah/integration/Makefile (hunk 3 above)
Comment (chamons) [reply tools/nnyeah/integration/Makefile:13]: Yep, I did that because they are broken (see https://github.com/xamarin/xamarin-macios/pull/14902).

File: tools/nnyeah/integration/Makefile (hunk 2 above)
Comment (chamons) [reply tools/nnyeah/integration/Makefile:4]: Good point. Let me fix that.

**Comment by chamons** [comment]
Moved to a template Makefile and standardized the csproj and assembly names.

File: tools/nnyeah/integration/Makefile [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+$(eval $(call INTEGRATION_TEST_template,tvOS))
+
+all-local:: $(TEST_BUILD_TARGETS)
+install-local:: all-local
+
+prep-tests:
+	$(Q) rm -rf converted
+	$(Q) mkdir -p converted	
+
+run-tests:: prep-tests $(TEST_RUN_TARGETS)
```
Comment (rolfbjarne) [inline tools/nnyeah/integration/Makefile:31]: This runs into problems with parallel make, `prep-tests` could be executed at the same time (or even after) the rn-* targets.

A different approach would be to make the output platform-specific above:

    --output=Converted/$(1)/Integration.dll

and then you deleted that directory first:

    rm -rf Converted/$(1)

So something like this:

```make
run-$(1): $(1)/bin/Debug/$(1)Integration.dll
 	$(Q) rm -rf Converted/$(1)
	$(Q) mkdir -p Converted/$(1)
 	-$(Q) $(SYSTEM_DOTNET) run --project ../nnyeah/nnyeah.csproj --input=$(1)/bin/Debug/$(1)Integration.dll --output=Converted/$(1)/Integration.dll
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tools/nnyeah/integration/MyClass.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+namespace MacIntegration
+{
+    public class MyClass
+    {
+        public MyClass()
```
Comment (tj-devel709) [inline tools/nnyeah/integration/MyClass.cs:6]:
```suggestion
        public MyClass ()
```

File: tools/nnyeah/integration/MyClass.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+        {
+        }
+
+        public nint Echo(nint x) => x;
```
Comment (tj-devel709) [inline tools/nnyeah/integration/MyClass.cs:10]:
```suggestion
        public nint Echo (nint x) => x;
```

**Review by tj-devel709 — APPROVED** [review]
nit - style but everything else LGTM as far as I can tell

File: tools/nnyeah/integration/Makefile (hunk 4 above)
Comment (chamons) [reply tools/nnyeah/integration/Makefile]: Fixed!

**Review by rolfbjarne — APPROVED** [review]
